shiroclient/health.go

```diff
@@ -0,0 +1,146 @@
+package shiroclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+)
+
+type jsonFieldError struct {
+	desc  string
+	typ   string
+	field string
+}
+
+func (err *jsonFieldError) Error() string {
+	if err.typ == "" {
+		return fmt.Sprintf("%s expected %s field", err.desc, err.field)
+	}
+	return fmt.Sprintf("%s expected %s %s field", err.desc, err.typ, err.field)
+}
+
+func stringFieldError(desc string, field string) *jsonFieldError {
+	return &jsonFieldError{desc, "string", field}
+}
+
+type healthcheck []HealthCheckReport
+
+func (c healthcheck) Reports() []HealthCheckReport {
+	return c
+}
+
+type healthreport struct {
+	timestamp      string
+	status         string
+	servicename    string
+	serviceversion string
+}
+
+func (h *healthreport) Timestamp() string      { return h.timestamp }
+func (h *healthreport) Status() string         { return h.status }
+func (h *healthreport) ServiceName() string    { return h.servicename }
+func (h *healthreport) ServiceVersion() string { return h.serviceversion }
+
+var _ HealthCheckReport = (*healthreport)(nil)
+
+// NOTE:  convertHealthReport doesn't unmarshal directly into the healthreport
+// struct to maintain semantics similar to other json decoding happening in
+// this package (e.g. semantics around handling incorrect letter cases and
+// missing fields).
+func unmarshalHealthResponse(r []byte) (healthcheck, error) {
+	// NOTE: rawResp *does* use json struct deserialization to ease handling of
+	// any exception object which may be passed from upstream.
+	var rawResp struct {
+		Reports   []interface{}
+		Exception *json.RawMessage
+	}
+	err := json.Unmarshal(r, &rawResp)
+	if err != nil {
+		return nil, fmt.Errorf("invalid result format: %w", err)
+	}
+	if rawResp.Exception != nil {
+		return nil, fmt.Errorf("remote exception: %s", *rawResp.Exception)
+	}
+	reports := make(healthcheck, len(rawResp.Reports))
+	for i, rawReport := range rawResp.Reports {
+		reports[i], err = convertHealthReport(rawReport)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return reports, nil
+}
+
+func convertHealthReport(rawReport interface{}) (*healthreport, error) {
+	m, ok := rawReport.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("health check report: expected an object")
+	}
+	const errdesc = "health check report"
+	ts, ok := m["timestamp"].(string)
+	if !ok {
+		return nil, stringFieldError(errdesc, "timestamp")
+	}
+	status, ok := m["status"].(string)
+	if !ok {
+		return nil, stringFieldError(errdesc, "status")
+	}
+	svc, ok := m["service_name"].(string)
+	if !ok {
+		return nil, stringFieldError(errdesc, "service_name")
+	}
+	ver, ok := m["service_version"].(string)
+	if !ok {
+		return nil, stringFieldError(errdesc, "service_version")
+	}
+	report := &healthreport{
+		timestamp:      ts,
+		status:         status,
+		servicename:    svc,
+		serviceversion: ver,
+	}
+	return report, nil
+}
+
+// smartHealthCheck is an internal interface that is not intended to be used in
```

Comment:
I think we're going to want to make a public interface at some point. That can probably wait until this is a little more stable, and might be at a higher level of abstraction than this, but I think we _will_ want it to be a legible ?subtype? of the ShiroClient interface.